compiler/rustc_infer/src/infer/relate/type_relating.rs

```diff
@@ -0,0 +1,317 @@
+use super::combine::CombineFields;
+use crate::infer::{DefineOpaqueTypes, ObligationEmittingRelation, SubregionOrigin};
+use crate::traits::{Obligation, PredicateObligations};
+
+use rustc_middle::ty::relate::{Cause, Relate, RelateResult, TypeRelation};
+use rustc_middle::ty::visit::TypeVisitableExt;
+use rustc_middle::ty::TyVar;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_span::Span;
+use std::mem;
+
+/// Enforce that `a` is equal to or a subtype of `b`.
+pub struct TypeRelating<'combine, 'a, 'tcx> {
+    fields: &'combine mut CombineFields<'a, 'tcx>,
+    a_is_expected: bool,
```

Comment:
`a_is_expected` should be unnecessary now, as we can instead flip the related types by using `xform(ty::Contravariance)`. Where do we use `a_is_expected` right now?

Reply:
It's not that we modify `a_is_expected` from within the relation. It's that we construct the relation (and therefore pass in `a_is_expected`) separately from where we use it. I think this is prohibitively invasive to fix right now.

Reply:
hmm, we could then simpy flip the args if `Trace { a_is_expected }` is false. But yeah, am fine with doing that in a separate PR